### PR TITLE
[CBRD-20055] updating incompatible timezone data to upgrade timezone library

### PR DESCRIPTION
**Pull request by asturzu**
http://jira.cubrid.org/browse/CBRD-20055
    In some situations the extend feature previously developed is not enough.
    In order to ensure backward compatibility with timezone data inserted in tables.
    Examples include: adding a new geographical timezone, modifying offset rules or daylight saving rules.
    In order to resolve this situation an update of the data in the tables that contain timezone
data types is required. The solution uses a function that converts the data in the tables using the 
current timezone library and the newly generated one. 
     The function first updates the timezone id which represents an index in the names array which
contains all the names of all the timezones and the aliases. In order to do the update
it uses the zone id from the old library to get the name of the timezone and after that
it does a binary search in the names array of the new library generated with the extend tool and
updates the zone id. 
     The second step of the conversion function is to update the offset rule id
and the daylight saving time rule id.  The offset rule id represents an index in the offset_rules array.
More exactly, it is used to retrieve the exact offset rule which applies to the current date.
The daylight saving rule id is an offset in the ds_rules array. It is used to retrieve the exact daylight
saving time rule which applies to the current date.
     The time stored on disk is in UTC time reference.  Using the two ids mentioned above which
are also stored on disk the time is transformed in local time using the current timezone library.
Then, using the new timezone library it is transformed back in UTC time reference. Also, the offset rule id
and the daylight saving time id are also updated. The transformation is done using the tz_datetime_utc_conv function which can go from UTC time reference to local and the oposite.
     After doing this conversion the data is made backward compatible with the new generated timezone
library. 
    In order to update all the tables an update query is used which calls the conversion function.
   More exactly, it does the following:
       update table_name set column_name = conv_tz (column_name);
   In order to execute the query, the grammar has been modified by adding the conv_tz function.
  The conv_tz function returns the same data type as the data type given as argument. More exactly,
it can take as arguments the following data types:
    DATETIMETZ, DATETIMELTZ, TIMESTAMPTZ, TIMESTAMPLTZ. 

   In order to retrieve all the tables in a database a show tables query was used. Also, the data
is migrated for all the databases stored on disk.

### CODE REVIEW

**Comment by eseokoh** [comment]
@asturzu 

Please add descriptions to your fix. 
Thank you.

**Comment by eseokoh** [comment]
@asturzu 

Please let me give you my first feedback. 

I expect only few of rules are changed. (New ones are appended at the end and given unused ids.)
Since the current update query will update the entire records of database which includes a timezone type, I'd like to minimize the affected rows. 
Please imagine something like this: `update t set c1 = conv_tz (c1) where c1 != conv_tz (c1);`.
I don't this is the best available query for us. We have to think about it. 

Another optimization I'd like to propose is to update multiple columns of a record as an update. Let's say a table has multiple TZ columns. We can issue update query: `update t set c1 = conv_tz(c1), c2 = conv_tz(c2) ... where c1 != conv_tz (c1) or c2 != conv_tz (c2) ... ;`.

One more issue is "upgrade scenario", especially for HA environment. Upgrading a single node can be done as SA mode. However, doing HA nodes is not easy as single mode.

I'll return to you again with more observations.

**Comment by asturzu** [comment]
Dear Mr Oh,

I will implement your observations regarding updating columns that have timezone data types.

     Antonio

**Comment by asturzu** [comment]
Dear Mr Oh,

I think that using this update t set c1 = conv_tz (c1) where c1 != conv_tz (c1)
won't help that much, because you would still call the conv_tz function which does all the 
important things and you would still need to do a compare. I think that I could speed up things
for timezones which have not been affected by the update by speeding up the conv_tz function.
I think I could speed it up by doing an additional preprocessing in the tzc_extend function and 
storing in a global array for each timezone id if it is backward compatible or not. Using this I
could return immediately from the conv_tz function for the timezones that have not been changed.

**Comment by eseokoh** [comment]
The idea sounds good. I think we can merge two ideas. :)
One concern of the first version of update is the amount of transaction logs that will write. 
Let's say a query:
```
update t set a = a, b = b;
```
It will update all records and surely write all undo/redo logs. However, it does nothing from the point of data. 

I expect the most of timezone data remains same as before and almost all undo/redo logs are actually meaningless. 

We may consider to skip update a record when its new value is same as its current. (I believe this is the behavior of MySQL.) This is another issue. 

When we have the updated version of conv_tz function, the query will scan tables which has a timezone data type, promptly filter the affected records and update them. Even though computation of the function takes time, it will be acceptable since the affected rows are small for many cases.

So, I'd like to keep proposing the query:
```
update t set c1 = conv_tz(c1), c2 = conv_tz(2) where c1 != conv_tz(c1) or c2 != conv_tz(c2);
```

Thank you.

**Comment by eseokoh** [comment]
Another idea come up with me is to use a session variable to minimize evaluation of tz_conv. 

```
update t1 set c1 = @s1, c2 = @s2 where t1.c1 != (@s1 := tz_conv (c1)) or t1.c2 != (@s2 := tz_conv (c2));
```

**Comment by eseokoh** [comment]
sorry, no. 
It will not work, since the latter predicates of where clause will not be evaluated when a former one is satisfied. Session variables for the latter parts are remained as old values. 
It can be applied to tables which has only one timezone column, or we have to issue an update query for a timezone column.

**Comment by eseokoh** [comment]
Since I think cost of scan is not comparable with that of function evaluation and I don't also want to separate logic, I'd like to return to the query:
```
update t set c1 = conv_tz(c1), c2 = conv_tz(c2) where c1 != conv_tz(c1) or c2 != conv_tz(c2);
```

**Comment by eseokoh** [comment]
@asturzu 

I've pushed a few changes. Please confirm it. 
I also left FIXME at tzc_update and hope you will find nice idea and fill it. :)

Please also confirm usages of `gen_tz`:
```
usage: cubrid gen_tz [OPTION]
       cubrid gen_tz -g extend [OPTION] [database-name]
```
I think it should be also fixed.

**Comment by asturzu** [comment]
@eseokoh 

I read your patch. There was compilation and linking error due to the fact
that you made tz_Is_backward_compatible_timezone, tz_Compare_datetimetz_tz_id, tz_Compare_timestamptz_tz_id SA_MODE. To fix the errors I had to make tzc_extend, tzc_update and
timezone_compile_data also SA_MODE. Also, I moved db_query_end(result3) after the error checking to avoid evaluating something that might be NULL. The rest of the patch seems correct.
An observation could be that most of the functions in tz_compile.c could be made also only SA_MODE.
I will update the help for gen_tz on Linux because it seems out of date.

**Comment by asturzu** [comment]
@eseokoh 

I modified the utils.msg file in order to update the help message for gen_tz but excluding English
and Romanian somebody else will have to translate them in the rest of the languages.

File: src/base/tz_compile.c [hunk 1]
```diff
@@ -6720,11 +6720,11 @@ tzc_update (TZ_DATA * tzd, const char *database_name)
 			}
 		      db_query_end (result3);
 		    }
-		  printf("Finished updating table %s\n", table_name);
+		  printf ("Finished updating table %s\n", table_name);
 		}
 	    }
 	}
-      printf("Finished updating database %s\n", db_info_p->name);
+      printf ("Finished updating database %s\n", db_info_p->name);
```
Comment (eseokoh) [inline src/base/tz_compile.c]: please move it after `commit` and also remove `FIXME` comment. :)

**Comment by asturzu** [comment]
@eseokoh 

I have a question about the merge. I didn't include csql_grammar.c, csql_grammar.h and csql_lexer.c
files because they are generated using csql_grammar.y file. I didn't include them because I was
afraid of merge conflicts.  Should I include them also ?

**Comment by eseokoh** [comment]
@asturzu 

Yes, please. 
It helps Windows build.

When you have a merge conflict, you are not required to manually resolve it. You can just resolve the conflict, then touch .y file and build it. It will generate new .[ch] and you can merge it.

**Pull request merged**